src/confluence.ml

```diff
@@ -0,0 +1,78 @@
+(** First-order syntactic unification *)
+
+open Terms
+open Basics
+open Timed
+
+exception Unsolvable
+
+(** [unif_aux t1 t2 eqs] attempts to solve the first-order unification
+    problem [(t1 = t2) :: eqs]. *)
+let rec unif_aux : term -> term -> unif_constrs -> unit = fun t1 t2 eqs ->
+  match (t1, t2) with
+  | Type, Type
+  | Kind, Kind                    -> unif eqs
+  | Symb(s1, _), Symb(s2, _)      ->
+      if s1 == s2 then unif eqs else raise Unsolvable
+  | Vari _, _
+  | _, Vari _
+  | Prod _, _
+  | _, Prod _
+  | Abst _, _
+  | _, Abst _                     -> assert false
+  | Appl (t1, u1), Appl (t2, u2)  -> unif ((t1, t2) :: (u1, u2) :: eqs)
+  | Meta (m1, ts1), Meta (_, ts2) ->
+      assert (ts1 = [||] && ts2 = [||]);
+      let vs = Bindlib.new_mvar mkfree [||] in
+      let bt2 = Bindlib.bind_mvar vs (lift t2) in
+      set_meta m1 (Bindlib.unbox bt2);
+      unif eqs
+  | (Meta (m1, ts1), t2)
+  | (t2, Meta (m1, ts1))          ->
+      if occurs m1 t2 then raise Unsolvable
+      else
+        assert (ts1 = [||]);
+        let vs = Bindlib.new_mvar mkfree [||] in
+        let bt2 = Bindlib.bind_mvar vs (lift t2) in
+        set_meta m1 (Bindlib.unbox bt2);
```

Comment:
Can't this be simplified/optimized @rlepigre ?

Reply:
@rlepigre : In addition, as references are timed, and all these meta and meta substitutions are only used for checking subject reduction, shouldn't we through all these references away after subject reduction using save/restore to limit memory usage? Idem for TRef instantiations used in eq when calling the tactic rewrite.

Reply:
We could maybe avoid a lifting/unboxing here. Is the term `t2` guaranteed to be closed here?

For the second point, I don't think this would make much difference. The memory is collected for every undo that is not reachable from a reachable saved time, so I would not worry too much about that.